File: 2018/python/day02.py
def partA(boxes):
    twoCount = 0
    threeCount = 0
    for box in boxes:
        boxHasTwoLetters = False
        boxHasThreeLetters = False
        letters = {}
        for char in box:
            if char in letters:
                letters[char] += 1
            else:
                letters[char] = 1
        for char, num in letters.items():
            if num == 2 and not boxHasTwoLetters:
                boxHasTwoLetters = True
                twoCount += 1
            if num == 3 and not boxHasThreeLetters:
                boxHasThreeLetters = True
                threeCount += 1
            if boxHasThreeLetters and boxHasTwoLetters:
                break
    return twoCount * threeCount

def partB(boxes):
    allBoxSubsets = set()
    for box in boxes:
        boxSubsets = set()
        for i, _ in enumerate(box):
            boxSubsets.add(box[:i] + box[i+1:])
        answerSet = allBoxSubsets & boxSubsets
        if len(answerSet) == 1:
            return list(answerSet)[0]
        allBoxSubsets.update(boxSubsets)

def main():
    file = "input.txt"
    boxes = []
    with open(file) as f:
        for line in f:
            boxes.append(line.strip())
    print("Solving Day02A...")
    a = partA(boxes)
    print(a)
    print("Solving Day02B...")
    b = partB(boxes)
    print(b)

main()

File: 2018/ts/src/day02.ts
import { readFileSync } from "node:fs";

type Boxes = Array<string>;
type Checksum = number;

export const readInput = (file: string = "./src/input.txt"): Boxes => {
  const data = readFileSync(file);
  const str = data.toString();
  return str.split("\n");
}

export const partA = (boxes: Boxes): Checksum => {
  // keep number of 2 and 3 times occurances for all boxes
  const occurances = { 2: 0, 3: 0 };
  for (const box of boxes) {
    // count letter frequency for each box
    const lettersCount: Record<string, number> = {};
    for (const c of box) {
      if (c in lettersCount) {
        lettersCount[c] += 1;
      } else {
        lettersCount[c] = 1;
      }
    }
    let countTwo = false;
    let countThree = false;
    // increase 2 and 3 times occurances counter as necessary
    for (const [k, v] of Object.entries(lettersCount)) {
      if (v === 3 && !countThree) {
        countThree = true;
        occurances[3] += 1;
      }
      else if (v === 2 && !countTwo) {
        countTwo = true;
        occurances[2] += 1;
      }
      if (countTwo && countThree) {
        // we do not care about repetitions of two and three times occurances per single box
        break;
      }
    }
  }
  return occurances[2] * occurances[3];
}

export const partB = (boxes: Boxes): string | null => {
  const oneLetterLessStrings: Set<string> = new Set();
  for (const box of boxes) {
    const letters = box.split("");
    // we do not want duplicate sequences from the same box
    const boxOneLessLetterSet = new Set<string>();
    for (let i = 0; i < letters.length; i++) {
      let lettersCopy = [...letters];
      lettersCopy.splice(i, 1);
      const oneLessLetters = lettersCopy.join("");
      boxOneLessLetterSet.add(oneLessLetters);
    }
    for (const str of boxOneLessLetterSet) {
      // as soon as we encounter the same key we are done
      if (oneLetterLessStrings.has(str)) {
        return str;
      }
      oneLetterLessStrings.add(str);
    }
  }
  return null;
}

export const main = () => {
  const boxes = readInput();
  console.log("Solving part A...");
  const a = partA(boxes);
  console.log(a);
  console.log("Solving part B...");
  const b = partB(boxes);
  console.log(b);
};
